oak_idl/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! A Rust-based IDL implemented via `macro_rules`. It is inspired by
+//! <https://github.com/google/tarpc> and
+//! <https://fuchsia.dev/fuchsia-src/reference/fidl/language/wire-format>.
+//!
+//! For each service definition, it generates both a client stub and a service trait, as well as a
+//! method to dispatch invocations to a service implementation.
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+pub enum ClientError {
+    InvalidRequest,
+    InvalidResponse,
+    TransportError(TransportError),
+}
+
+impl From<TransportError> for ClientError {
+    fn from(e: TransportError) -> Self {
+        ClientError::TransportError(e)
+    }
+}
+
+#[derive(Debug)]
+pub enum TransportError {
+    InvalidMessage,
+    InvalidMethodId,
+}
+
+/// Unique identifier of a method within a service.
+type MethodId = u32;
+
+/// Header of a transactional message.
+///
+/// See <https://fuchsia.dev/fuchsia-src/reference/fidl/language/wire-format#transactional-messages>
+#[derive(Serialize, Deserialize)]
+pub struct Header {
+    pub transaction_id: u32,
```

Comment:
Can you clarify what the transaction_id does? Is it supposed to be globally unique, unique in combination with the method id?